### ISSUE TITLE
Consider using Shadow Collection?

### ISSUE THREAD

**Issue by Kestanous** [dburles/meteor-collection-helpers#26]
I love this package but it had a few issues and think I have build a solution to them with: https://github.com/Meteor-Reaction/shadow-collection. 

What you would get to using this:
* non-conflict - other packages that want to proved helpers can do so without conflicting with this package.
* it lets Mongo.Collection transform work again
* It removes the console reference to `Mongo.Collection.helpers.Document.(anonymous function)`

If this interests you I would be happy to do a PR, or leave it to you.

**Comment by Kestanous** [comment]
Also worth noting that I have not, yet, published shadow-collection as I am willing to make adjustments as needed.

**Comment by dburles** [comment]
Hey @Kestanous interesting idea, I'm interested to know more about your use-case for this. I can see how it might be a nice idea to extend helpers from an existing transform, but on the other hand I worry that this could easily lead to very difficult bugs to trace. For example if you were to unknowingly override an existing method.

**Comment by Kestanous** [comment]
@dburles My end goal is full models: https://github.com/Meteor-Reaction/gravity-models. But @rclai work inspired this code https://github.com/rclai/meteor-document-methods.

Maybe we could add a warning or throw an exception if an existing method were to be overwritten? I am not sure just yet but I think the idea is worth investigating. Happy for feedback, actually if you are willing to switch to this I can just add you to the "shadow" teem ([shadow-collection](https://github.com/Meteor-Reaction/shadow-collection) & [shadow-describer](https://github.com/Meteor-Reaction/shadow-describer)). That way you can you can maintain collection-helpers without delays.

**Comment by dburles** [comment]
Perhaps an idea might be to provide a configuration option that will allow this behaviour. By default it'll throw an error.

**Comment by Kestanous** [comment]
@dburles thats also what I was thinking. It would be neat if we could add an exception rule for each method, rather then all or nothing.

```
exception = {
  helpers: false,  //throw error
  sillyName: true //allow override
}
```

**Comment by rhyslbw** [comment]
:+1:

**Comment by dburles** [comment]
I'm not totally sold on the idea of the shadow collection package, I feel there's not a real-world requirement for that functionality. I'm interested to hear though if it was built within the context of a real-world application and how having that functionality was required.

**Comment by Kestanous** [comment]
@dburles so there is a real-world need but it is very abstract at this level. 

The issue I often have is I need my model layer to be self describing and self encapsulating. Your packages (helpers & instances) does a really go job on the encapsulating side. I am building [shadow-describer](https://github.com/Meteor-Reaction/shadow-describer) to cover the description side.

Consider the following (obviously the code could be cleaned up)
Here is how it is now:
```js
dynamicAddToFoo = function (data, collection, type) {
  if (type === 'cursor') {
    data.foreach(function (obj) {
      Mongo.Collection.get(collection).update(obj._id, { foo: ++obj.foo });
    })
  } else {
    Mongo.Collection.get(collection).update(data._id, { foo: ++data.foo });
  }
}
```

vs
```js
dynamicAddToFoo = function (data) {
  if (data.$shadow.type === 'cursor') {
    data.foreach(function (obj) {
      Mongo.Collection.get(data.$shadow.collectionName).update(obj._id, { foo: ++obj.foo });
    })
  } else {
    Mongo.Collection.get(data.$shadow.collectionName).update(data._id, { foo: ++obj.foo });
  }
}
```

Now imagine the above code is in a package, say a form package, where the collection name may not be known. This adds a whole different level of abstraction to our data. In order to get [shadow-describer](https://github.com/Meteor-Reaction/shadow-describer), [validity]( https://github.com/Meteor-Reaction/meteor-validity), and [document-methods](https://github.com/rclai/meteor-document-methods) to play nice with collection-helpers, I need it to use shadow-collections. 

tl;dr
I think it makes sense to add this but if you still disagree then no worries. I will likely just end up creating a shadow-helpers package that suits my needs. Mostly I think this can help the community as a whole and not just my apps.

**Comment by Kestanous** [comment]
Actually I forgot I was adding a pointer to the collection itself in shadow-describer so I would not actually need to use your collection-instances package.

**Comment by dburles** [comment]
I hope you can understand that I'm a bit hesitant to make changes such as making use of shadow-helpers, as it's doing some funky stuff, considering the number of people already using this package (+ production apps). I'll have a bit more of a think about the idea of extending from an existing transform.

**Comment by dburles** [comment]
Just as a side note, it's possible for a package to depend on collection-helpers and apply helpers just as normal. The `Collection.helpers` function can be called as many times as you like. So you can possibly achieve things that way.

**Comment by Kestanous** [comment]
@dburles I fully understand and appreciate your hesitation, this needs to be done right or not at all. Here are my thoughts.

1: Can you speak to the part that bothers you the most? I don't see how this is much worse then what is already happening and its mutates much less then Collection2 or PeerDB. Do you see a way we could make this safe enough to make it work. I am willing to do what it takes to get us there. If you could elaborate on the risks you see I will do my best to nullify them.

2: Unfortunately your solution would not solve the following issues I am facing, but maybe you see an alternative to those as well.

I need each instance, output from a cursor, to have a hidden object for state management. I am looking at using something like [observe-js](https://github.com/Polymer/observe-js) to watch changes on an object, auto run validations live, and keep a change log. To do that I need to store that state, for each instance, outside of the instance itself. 

[validity](https://github.com/Meteor-Reaction/meteor-validity) is checking if a schema exists on the collection *before* it adds the `$validate` helper. This prevents polluting collections that done need validations and could also let apps define schemas after meteor startup. I don't see a real world need for the second option but there you have it.

@rclai's [document-methods](https://github.com/rclai/meteor-document-methods) is using a nasty hack to play nice with collection-helpers. I would like to remove that need. (@rclai it was very well done though)

**Comment by Kestanous** [comment]
I am willing to look into alternatives for all of that but shadow-* packages solve so many of my issue. Both in production and in my head -_^

**Comment by rclai** [comment]
There was a point when my package depended on `collection-instances` and looked really nice: https://github.com/rclai/meteor-document-methods/blob/471f583ddf2037751a38acf1fad8528a5f10ddc6/document-methods.js

The reason I moved into the monkey patching technique was because collections are usually instantiated outside of a `Meteor.startup` in a `lib` folder and some people might end up doing some collection operations outside of `Meteor.startup` and if they do that, the methods are undefined because the runtime hasn't reached the point where I inject the methods. So essentially I wanted the methods to be available at the earliest time possible and not just after `Meteor.startup`.

Should I should go back to what I had?

**Comment by dburles** [comment]
Hey @rclai why were you calling `Mongo.Collection.getAll()` within startup?

**Comment by Kestanous** [comment]
@dburles would getting more opinions on this help? We could to a community feedback request on crater be of value? We could link to this issue or you could lock this and we could move this to crater.

**Comment by dburles** [comment]
I think I'll park on it for a while.

**Comment by Kestanous** [comment]
@dburles just so you know I have a temp package over at https://github.com/Meteor-Reaction/shadow-helpers.

**Comment by rhyslbw** [comment]
@Kestanous :+1: This is a good fit for my needs too

**Comment by rclai** [comment]
When adding transforms on top of another, isn't it going to cause the documents to no longer share a prototype function but rather all have a copy of the function at the top level?

**Comment by Kestanous** [comment]
@rclai no because it defines the functions for each doc, see https://github.com/Meteor-Reaction/shadow-collections/blob/master/lib/shadow-collection.js#L35.
